TESTS/mbed_hal/device_key/main.cpp

```diff
@@ -0,0 +1,126 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+#include "mbed.h"
+#include "device_key_api.h"
+
+#ifndef DEVICE_DEVKEY
+#error [NOT_SUPPORTED] DEVICE_KEY needs to be enabled for this test
+#endif
+
+using namespace utest::v1;
+
+enum TestDeviceKeySize {
+
+    DEVICE_KEY_16BYTE = 16,
+    DEVICE_KEY_32BYTE = 32
+};
+
+/*
+ * Test if the platform return a valid device key size
+ */
+void device_key_get_size_test()
+{
+    size_t len = device_key_get_size_in_bytes();
+    TEST_ASSERT_FALSE_MESSAGE(DEVICE_KEY_16BYTE != len && DEVICE_KEY_32BYTE != len,
+            "Device key length is not 16 or 32 Byte long");
+}
+
+/*
+ * Test that the platform return error when the buffer is to small
+ */
+void device_key_get_key_buffer_to_small_test()
+{
+    uint32_t buffer[(DEVICE_KEY_16BYTE/2) / sizeof(uint32_t)];
+    size_t len = DEVICE_KEY_16BYTE/2;
+
+    memset(buffer, 0, DEVICE_KEY_16BYTE/2);
+    int status = device_key_get_value(buffer, &len);
+    TEST_ASSERT_EQUAL_INT32(DEVICEKEY_HAL_BUFFER_TOO_SMALL, status);
+}
+
+/*
+ * Test if the platform return a device key buffer with
+ * the correct length.
+ */
+void device_key_get_key_length_test()
+{
+    size_t expected = device_key_get_size_in_bytes();
+    uint32_t buffer[(DEVICE_KEY_32BYTE*2) / sizeof(uint32_t)];
+    size_t len = (DEVICE_KEY_32BYTE*2);
+
+    memset(buffer, 0, DEVICE_KEY_32BYTE*2);
```

Comment:
It's better to use `sizeof(buffer)` as the size (in case you decide to change the buffer size.